uml4net.xmireader/Cache/ExternalReferenceResolver.cs

```diff
@@ -0,0 +1,164 @@
+    /// <param name="logger">The logger for logging information and errors.</param>
+    public class ExternalReferenceResolver(IXmiReaderCache cache, HttpClient httpClient, IXmiReaderSettings settings, ILogger<ExternalReferenceResolver> logger)
+        : IExternalReferenceResolver
+    {
+        /// <summary>
+        /// Asynchronously attempts to resolve external references and yields their context and stream.
+        /// </summary>
+        /// <returns>An asynchronous enumerable of tuples containing the context and stream of resolved references.</returns>
+        public async IAsyncEnumerable<(string Context, Stream Stream)> TryResolve()
+        {
+            foreach (var identifier in cache.Cache.Values
+                         .SelectMany(cacheEntry => cacheEntry.SingleValueReferencePropertyIdentifiers.Values))
+            {
+                if(await this.TryResolve(identifier) is { Stream: { Length: >0 } } reference)
+                {
+                    yield return reference;
+                }
+            }
+
+            foreach (var identifiers in cache.Cache.Values
+                         .SelectMany(cacheEntry => cacheEntry.MultiValueReferencePropertyIdentifiers.Values))
+            {
+                foreach (var identifier in identifiers)
+                {
+                    if (await this.TryResolve(identifier) is { Stream: { Length: > 0 } } reference)
+                    {
+                        yield return reference;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously attempts to resolve an external reference identified by the specified key.
+        /// </summary>
+        /// <param name="key">The key representing the external reference to resolve.</param>
+        /// <returns>A task that represents the asynchronous operation, containing a tuple with the context and stream if resolved; otherwise, a default value.</returns>
+        private async Task<(string Context, Stream Stream)> TryResolve(string key)
+        {
+            if (string.IsNullOrEmpty(key) || key.StartsWith('#') || !key.Contains('#') || 
+                !cache.TryResolveContext(key, out var resource))
+            {
+                logger.LogInformation("Invalid external resource key [{key}]", key);
+                return default;
+            }
+
+            if (cache.DoesContextExists(resource.Context, resource.ResourceId))
+            {
+                logger.LogInformation("The resource {resource} was already parsed", resource.Context);
+                return default;
+            }
+            
+            try
+            {
+                var stream = default(Stream);
+
+                if (Uri.TryCreate(key, UriKind.Absolute, out var uri))
+                {
+                    if (uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps)
+                    {
+                        stream = await this.FetchRemoteXmi(uri);
```

Comment:
this i find tricky. lets discuss whether we every want to fetch an XMI document from the internet. I don't think we should. My proposal would be to then make sure it is avialable locally (make sure its safe to be processed) and make sure that the resource is mapped to a local resource by providing an override somehow